pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -44,115 +51,208 @@ func (o overlappingStatefulSets) Less(i, j int) bool {
+	claims := getPersistentVolumeClaims(set, pod)
+	newVolumes := make([]v1.Volume, 0, len(claims))
+	for name, claim := range claims {
+		newVolumes = append(newVolumes, v1.Volume{
+			Name: name,
+			VolumeSource: v1.VolumeSource{
+				PersistentVolumeClaim: &v1.PersistentVolumeClaimVolumeSource{
+					ClaimName: claim.Name,
+					// TODO: Use source definition to set this value when we have one.
+					ReadOnly: false,
+				},
+			},
+		})
 	}
-	blockingPetPod := blockingPet.pod
-	if hc.isHealthy(blockingPetPod) && !hc.isDying(blockingPetPod) {
-		glog.V(4).Infof("Clearing blocking pod %v for StatefulSet %v because it's healthy", bp.pod.Name, ps.Name)
-		u.store.Delete(blockingPet)
-		blockingPet = nil
+	for i := range currentVolumes {
+		if _, ok := claims[currentVolumes[i].Name]; !ok {
+			newVolumes = append(newVolumes, currentVolumes[i])
+		}
 	}
-	return blockingPet, nil
+	pod.Spec.Volumes = newVolumes
 }
 
-// Add records the given pet as a blocking pet.
-func (u *unhealthyPetTracker) Add(blockingPet *pcb) error {
-	u.storeLock.Lock()
-	defer u.storeLock.Unlock()
+// updateIdentity updates pod's name, hostname, and subdomain to conform to set's name and headless service.
+func updateIdentity(set *apps.StatefulSet, pod *v1.Pod) {
+	pod.Name = getPodName(set, getOrdinal(pod))
+	pod.Namespace = set.Namespace
+	if pod.Annotations == nil {
+		pod.Annotations = make(map[string]string)
+	}
+	pod.Annotations[podapi.PodHostnameAnnotation] = pod.Name
+	pod.Annotations[podapi.PodSubdomainAnnotation] = set.Spec.ServiceName
+}
 
-	if blockingPet == nil {
-		return nil
+// isRunningAndReady returns true if pod is in the PodRunning Phase, if it has a condition of PodReady, and if the init
+// annotation has not explicitly disabled the Pod from being ready.
+func isRunningAndReady(pod *v1.Pod) bool {
+	if pod.Status.Phase != v1.PodRunning {
+		return false
 	}
-	glog.V(4).Infof("Adding blocking pod %v for StatefulSet %v", blockingPet.pod.Name, blockingPet.parent.Name)
-	return u.store.Add(blockingPet)
+	podReady := v1.IsPodReady(pod)
+	// User may have specified a pod readiness override through a debug annotation.
+	initialized, ok := pod.Annotations[apps.StatefulSetInitAnnotation]
+	if ok {
+		if initAnnotation, err := strconv.ParseBool(initialized); err != nil {
+			glog.V(4).Infof("Failed to parse %v annotation on pod %v: %v",
+				apps.StatefulSetInitAnnotation, pod.Name, err)
+		} else if !initAnnotation {
+			glog.V(4).Infof("StatefulSet pod %v waiting on annotation %v", pod.Name,
+				apps.StatefulSetInitAnnotation)
+			podReady = initAnnotation
+		}
+	}
+	return podReady
 }
 
-// newUnHealthyPetTracker tracks unhealthy pets that block progress of statefulsets.
-func newUnHealthyPetTracker(pc petClient) *unhealthyPetTracker {
-	return &unhealthyPetTracker{pc: pc, store: cache.NewStore(pcbKeyFunc)}
+// isCreated returns true if pod has been created and is maintained by the API server
+func isCreated(pod *v1.Pod) bool {
+	return pod.Status.Phase != ""
```

Comment:
Note that we need to stop using Phase as much as possible: https://github.com/kubernetes/kubernetes/issues/7856